onnx/utils.py

```diff
@@ -21,3 +24,138 @@ def polish_model(model):  # type: (ModelProto) -> ModelProto
     model = onnx.optimizer.optimize(model)
     onnx.checker.check_model(model)
     return model
+
+
+class Extractor:
+    def __init__(self, model):  # type: ignore
+        self.model = onnx.shape_inference.infer_shapes(model)
+        self.graph = self.model.graph
+        self.wmap = self._buildNameDict(self.graph.initializer)
+        self.vimap = self._buildNameDict(self.graph.value_info)
+
+    @staticmethod
+    def _buildNameDict(objs):  # type: ignore
+        return {obj.name: obj for obj in objs}
+
+    def _filterTensors(self, originals, new_names):  # type: ignore
+        tmap = self._buildNameDict(originals)
+        original_names = set(tmap.keys())
+        s_new_names = set(new_names)
+        names_keep = s_new_names & original_names
+        names_add = s_new_names - original_names
+
+        tensors = []
+        for name in names_keep:
+            tensors.append(tmap[name])
+        for name in names_add:
+            # activation become input or output
+            tensors.append(self.vimap[name])
+
+        # sort the tensors
+        new_tmap = self._buildNameDict(tensors)
+        sorted_tensors = [new_tmap[name] for name in new_names]
+        return sorted_tensors
+
+    def _filterInputs(self, names):  # type: ignore
+        return self._filterTensors(self.graph.input, names)
+
+    def _filterOutputs(self, names):  # type: ignore
+        return self._filterTensors(self.graph.output, names)
+
+    def _dfsSearchReachableNodes(self, node_output_name, graph_input_names, reachable_nodes):  # type: ignore
```

Comment:
It seems like if a graph-input X is required to compute a specified-output Y, but if X is not in the specified-inputs, we could end up producing an incorrect model. May be it would help to check for this condition?

Reply:
Yes, the _missing input-X_ won't be taken as _input of the resulted model_. But I am not planning to detect such a corner case as it seems not to be easy work. Hope that the user may avoid this, and it can be caught by `onnx.check_model()` if we really encounter some...